### ISSUE TITLE
Deploy Keycloak to a test server

### ISSUE THREAD

**Issue by GerberI** [SiLeBAT/fcl-server#34]
https://gitlab.bfr.berlin/devs43/common/keycloak/issues/13

IT interaction is necessary

**Comment by afalenski** [comment]
VM works, I see the admin console, can login into the mibi portal and see the emails separately in another window.
When I register and then click on the link "Link zur Bestätigung der E-Mail-Adresse" in the email directly this window pops up:
![image.png](https://images.zenhubusercontent.com/5a854eea4b5806bc2bc70715/47e5d80d-494b-46b7-beb6-98704ea9a047)
Is this for the user or the admin? 
As you can see, the email is missing, options for user roles as well.

After registration and clicking "Ja" (see image above) no new entry was in the admin console at 'users'.

Currently, it is very hard to say that I could work with the present keycloak implementation. It feels "raw" and like several applications next to each other and not like one application. I need to try out a later implementation.

**Comment by GerberI** [comment]
@afalenski 
Thanks for testing.
- Yes it feels "raw", there are no customizations in the ui and messages implemented yet.
- Yes there are at least two applications in the user's view, the mibi-portal and the authentication server. This is on purpose because the authentication server is independent of the apps. For the daily work there is only the additional admin console for the user manager which is a replacement for the admin util we use now.
- The window "OAuth gewähren" is for the user. It is optional and can be removed. The purpose is to let the user decide what data an app is allowed to use. So it is kind of an data protection agreement per peace of data.
- There should be a user in the admin console after the user confirms the registration form. You have to refresh the user list to see the changes. E.g. select the search box and press "enter".

So, the testing focus for this first implementation should be on the user authentication flow and behavior of the mibi-portal. It is to extract a final draft how the fcl portal should exactly look like and behave in combination with the auth server. So things to test and decide are:
Login/Logout/Page Refresh with and without sample data loaded, with single or multiple tabs opened. Or session expiration (can be simulated by logging out a user through the admin console).
And as always to try to break the application. So try to be a very bad user and do stupid things :)
In short, first things to test are security, authentication flow and app behavior. Look and feel is for the final steps.

**Comment by afalenski** [comment]
Register
- When registering, a window "Bedingungen und Konditionen" appears. What kind of content is meant to appear here? Is it intended for the data protection confirmation?
- When declining (click on "Ablehnen") the "Bedingungen und Konditionen", the MiBi-Portal startpage appears and a toaster message "Authentication error". This is unexpected. As a user I would like to read a message like "Ohne Zustimmung zu [whatever "Bedingungen und Konditionen" is good for] kann kein Nutzer-Account eingerichtet werden." Then the buttons "Zurück zu 'Bedingungen und Konditionen'" and "Beenden" would be helpful in the authentication workflow.
- When using an email address which was used before to register, the email field is underlined in red and a text appears "E-Mail existiert bereits." I think some kind of data protection issue? In the productive MiBi-Portal the following toaster message appears: "Fehler beim Registrieren. Eine E-Mail mit weiteren Informationen wurde an [email-adress] gesendet. Wenn Sie keine E-Mail erhalten, wenden Siesich bitte direkt per E-Mail an uns: mibi-portal@bfr.bund.de"
- It is possible to use the password "x".

Login
- Login works
- I logged in as user A and uploaded data. Then I used "impersonate" for user A and it was possible to login at the same time and to upload a different dataset. Could there be issues leading to data loss when the user support (= product owner (PO)) is logged in into an account which is currently active in some other German federal state?

Page refresh
- Page Refresh with and without sample data loaded, with single or multiple tabs opened: all possible, no warning "Seite neu laden". The content is still there.

Session expiration
- A user who was logged out due to session expiration sees the info box "Sie wurden abgemeldet" and clicks on the button "Anmelden". Here, he or she has to enter both username/email address and password. In case of a session expiration where uploaded user data are still in the browser it would be convenient if he or she would find his/her email address already entered in the first field so that only the password needs to be entered.

Logout
- logout works, the warning appears that uploaded data will be lost.
- logout through admin console works, user gets window "Sie wurden abgemeldet"
- Login into MiBi-Portal with credentials, open second tab with MiBi-Portal (is logged in as the same user), logout in Tab 2 -> also in Tab 1 the user is logged out.

Keycloak admin console: 
- It would be convenient for the user support (= PO) if the columns could be sortet (users: username, email, last name; users/[user]/sessions: started, last access, clients)
- In Manage/Sessions not all or no sessions are shown. The user support (= PO) needs to click the button "show sessions" to see them. It would be helpful if this was done automatically when entering the sessions view.
- When the user support (=product owner) is logged in with his/her account (account A), it is not possible to impersonate as another user (account B) via keycloak. In the same browser another tab opens with the identical username (account A) which is already logged in. It is not possible to copy the link in Personal Info\Applikationen\mibiapp and thus it is not possible to open account B in another browser. [This is just a statement, no feature request.]

- When trying to send data there is an error. Toaster message: "Es gab einen Fehler beim Versenden der Datei an das MiBi-Portal."

I like the admin console more and more. The 'impersonate user' feature is great and very convenient.

**Comment by GerberI** [comment]
**Register**
- When registering, a window "Bedingungen und Konditionen" appears. **What kind of content is meant to appear here?** Is it intended for the data protection confirmation?
  - _Yes, this will be the data protection offer._
- When declining (click on "Ablehnen") the "Bedingungen und Konditionen", the MiBi-Portal startpage appears and a toaster message "Authentication error". This is **unexpected**. As a user I would like to read a message like "Ohne Zustimmung zu [whatever "Bedingungen und Konditionen" is good for] kann kein Nutzer-Account eingerichtet werden." Then the buttons "Zurück zu 'Bedingungen und Konditionen'" and "Beenden" would be helpful in the authentication workflow.
  - _Yes, it was kind of a problem to show a message at all. But adjusting the flow itself is a good idea. Another solution could be to use another button like: "Ablehnen und Zurück zum Mibi-Portal" or something. Ticket: https://gitlab.bfr.berlin/devs43/common/keycloak/issues/34_
- When using an email address which was used before to register, the email field is underlined in red and a text appears "E-Mail existiert bereits." I think this is a data **protection issue**. In the productive MiBi-Portal the following toaster message appears: "Fehler beim Registrieren. Eine E-Mail mit weiteren Informationen wurde an [email-adress] gesendet. Wenn Sie keine E-Mail erhalten, wenden Sie sich bitte direkt per E-Mail an uns: mibi-portal@bfr.bund.de"
  - _This could be a compromise between user experience and security. With the MiBi-Portal behavior the user does not know if he already is registered and has to trouble support for make thing clear. Perhaps we should ask ISB how to handle that._
- It is possible to use the password "x".
  - _https://gitlab.bfr.berlin/devs43/common/keycloak/issues/1_

**Login**
- I logged in as user A and uploaded data. Then I used "impersonate" for user A and it was possible to login at the same time and to upload a different dataset. **Could there be issues leading to data loss** when the user support (= product owner (PO)) is logged in into an account which is currently active in some other German federal state?
  - _No, the data is stored locally in the users browser and can not be viewed or manipulated by the PO._

**Page refresh**
- Page Refresh with and without sample data loaded, with single or multiple tabs opened: all possible, no warning "Seite neu laden". The content is still there.
  - _Keep in mind that there also is no warning if you close the tab. But you can reopen the tab and data is restored._

**Session expiration**
- A user who was logged out due to session expiration sees the info box "Sie wurden abgemeldet" and clicks on the button "Anmelden". Here, he or she has to enter both username/email address and password. In case of a session expiration where uploaded user data are still in the browser it would be convenient if he or she would find his/her **email address already entered in the first field** so that only the password needs to be entered.
  - _Yes, that is possible, but requires some thinking and code adjustments i had no time for yet. Ticket: https://gitlab.bfr.berlin/devs43/common/keycloak/issues/35_

**Logout**
- Login into MiBi-Portal with credentials, open second tab with MiBi-Portal (is logged in as the same user), logout in Tab 2 -> also in Tab 1 the user is logged out.
  - _Keep in mind that informing the user during logout events is not fully implemented yet. E.g. if a session naturally expires, the user is not informed immidiatly._

**Keycloak admin console**
- It would be convenient for the user support (= PO) if the **columns could be sorted** (users: username, email, last name; users/[user]/sessions: started, last access, clients)
- In Manage/Sessions not all or no sessions are shown. The user support (= PO) needs to click the button "**show sessions**" to see them. It would be helpful if a **refresh** was done automatically when entering the sessions view.
  - _The admin console is not perfect. It could be adjusted but i don't think it is a good idea to mess with the console itself._
- When the user support (=product owner) is logged in with his/her account (account A), it is not possible to impersonate as another user (account B) via keycloak. In the same browser another tab opens with the identical username (account A) which is already logged in. It is not possible to copy the link in Personal Info\Applikationen\mibiapp and thus it is not possible to open account B in another browser. [This is just a statement, no feature request.]
  - _This is because we use single sign on. Only one user can be logged in per browser. So to really test stuff like impersonation you have to use different browsers. In general i have to dive more into impersonation to fully understand the behavior and possibilities. Also i have to test it more intensive what would become time intensive. So i would suggest to just ignore the feature for now._

- When trying to **send data** there is an **error**. Toaster message: "Es gab einen Fehler beim Versenden der Datei an das MiBi-Portal."
  - _Sending data with the testuser should be possible. If you add a new user you have to manually add the institue attribute to the new user with a valid institute as the value. (That's because there is no institute selection in the registration form yet)._
